contracts/src/bridge/SequencerInbox.sol

```diff
@@ -198,19 +202,23 @@ contract SequencerInbox is DelegateCallAware, GasRefundEnabled, ISequencerInbox
         emit SequencerBatchData(sequenceNumber, data);
     }
 
-    function _reportBatchSpending(bytes32 dataHash, uint256 sequenceNumber) internal {
-        delayedInbox.submitBatchSpendingReportTransaction(msg.sender, dataHash, sequenceNumber);
-    }
-
-    function dasKeysetHashFromBatchData(bytes memory data) internal pure returns (bytes32) {
-        if (data.length < 33 || data[0] & 0x80 == 0) {
-            return bytes32(0);
+    modifier validateBatchData(bytes calldata data) {
+        uint256 fullDataLen = HEADER_LENGTH + data.length;
+        if (fullDataLen > MAX_DATA_SIZE) revert DataTooLarge(fullDataLen, MAX_DATA_SIZE);
+        if (data.length > 0 && (data[0] & DATA_AUTHENTICATED_FLAG) == DATA_AUTHENTICATED_FLAG) {
+            revert DataNotAuthenticated();
         }
-        bytes32 temp;
-        assembly {
-            temp := mload(add(data, 33))
+        // the first byte is used to identify the type of batch data
+        // das batches expect to have the type byte set, followed by the keyset (so they should have at least 33 bytes)
+        if (data.length < 33 || data[0] & 0x80 == 0) {
+            // not a DAS batch, so we don't need keyset validation
+            _;
+        } else {
+            // we skip the first byte, then read the next 32 bytes for the keyset
+            bytes32 dasKeysetHash = bytes32(data[1:33]);
+            if (!dasKeySetInfo[dasKeysetHash].isValidKeyset) revert NoSuchKeyset(dasKeysetHash);
+            _;
```

Comment:
I'm not sure exactly how solidity modifiers work, but does putting the _ in two places cause code duplication? We could put it underneath the if-else instead.